File: src/scanners/dockle.ts
import type { IScanner, ImageSource, ScannerResult } from '../types/index.js';
import { execWithTimeout, getToolVersion, isToolAvailable, parseJsonFile, writeFallbackResult, createTimedResult } from './utils.js';

const TIMEOUT_MS = 180_000;

function buildCommand(source: ImageSource, outputPath: string): string {
  switch (source.type) {
    case 'docker':
      return `dockle --format json --output "${outputPath}" "${source.ref}"`;
    case 'tar':
      return `dockle --input "${source.path}" --format json --output "${outputPath}"`;
    case 'registry':
      throw new Error('Dockle does not support direct registry scanning');
  }
}

export class DockleScanner implements IScanner {
  readonly name = 'dockle';

  async scan(source: ImageSource, outputPath: string): Promise<ScannerResult> {
    const start = Date.now();
    try {
      const command = buildCommand(source, outputPath);
      await execWithTimeout(command, TIMEOUT_MS);

      const data = await parseJsonFile(outputPath);
      return createTimedResult(this.name, true, start, { data });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.warn('Dockle scan failed:', message);
      await writeFallbackResult(outputPath, message);
      return createTimedResult(this.name, false, start, { error: message });
    }
  }

  async getVersion(): Promise<string> {
    return getToolVersion('dockle --version');
  }

  async isAvailable(): Promise<boolean> {
    return isToolAvailable('dockle');
  }

  supportsSource(source: ImageSource): boolean {
    return source.type !== 'registry';
  }
}

File: internal/scanner/dockle.go
package scanner

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

const dockleTimeoutMs = 180000

type DockleScanner struct{}

func (d *DockleScanner) Name() string { return "dockle" }

func (d *DockleScanner) Scan(ctx context.Context, source types.ImageSource, outputPath string) (*types.ScannerResult, error) {
	start := time.Now()

	if source.Type == "registry" {
		msg := "Dockle does not support direct registry scanning"
		durationMs := time.Since(start).Milliseconds()
		return &types.ScannerResult{Scanner: "dockle", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	cmd := d.buildCommand(source, outputPath)
	_, _, err := ExecWithTimeout(ctx, cmd, dockleTimeoutMs, nil)
	durationMs := time.Since(start).Milliseconds()

	if err != nil {
		msg := err.Error()
		fmt.Fprintf(os.Stderr, "Dockle scan failed: %s\n", msg)
		_ = WriteFallbackResult(outputPath, msg, nil)
		return &types.ScannerResult{Scanner: "dockle", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	var data interface{}
	if parseErr := ParseJSONFile(outputPath, &data); parseErr != nil {
		msg := parseErr.Error()
		return &types.ScannerResult{Scanner: "dockle", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	return &types.ScannerResult{Scanner: "dockle", Success: true, Data: data, DurationMs: durationMs}, nil
}

func (d *DockleScanner) buildCommand(source types.ImageSource, outputPath string) string {
	switch source.Type {
	case "tar":
		return fmt.Sprintf(`dockle --input "%s" --format json --output "%s"`, source.Path, outputPath)
	default:
		return fmt.Sprintf(`dockle --format json --output "%s" "%s"`, outputPath, source.Ref)
	}
}

func (d *DockleScanner) GetVersion() string {
	return GetToolVersion("dockle --version")
}

func (d *DockleScanner) IsAvailable() bool {
	return IsToolAvailable("dockle")
}

func (d *DockleScanner) SupportsSource(source types.ImageSource) bool {
	return source.Type != "registry"
}
